haystack/components/rankers/metadata_grouper.py

```diff
@@ -0,0 +1,190 @@
+
+        If no 'group_by' value is present in the document, the document is added to a list of documents without a group.
+
+        :param documents:
+        :returns:
+            A tuple with the following elements:
+            - document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+            - no_group: A list of documents without a 'group_by' value.
+            - ordered_keys: A list of 'group_by' values in the order they were encountered.
+        """
+
+        document_groups: Dict[str, List] = defaultdict(list)
+        ordered_keys: List = []
+        no_group: List = []
+
+        for document in documents:
+            if self.group_by in document.meta:
+                document_groups[str(document.meta[self.group_by])].append(document)
+                if str(document.meta[self.group_by]) not in ordered_keys:
+                    ordered_keys.append(str(document.meta[self.group_by]))
+            else:
+                no_group.append(document)
+
+        return document_groups, no_group, ordered_keys
+
+    def _buck_into_subgroups(self, document_groups: Dict[str, List]) -> Tuple[Dict[str, Dict], List]:
+        """
+        Buckets the documents within the groups based on the 'subgroup_by' value.
+
+        If no 'subgroup_by' value is present in the document, the document is added to a subgroup with the key
+        'no_subgroup'.
+
+        :param document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+        :returns:
+            A tuple with the following elements:
+            - document_subgroups: A dictionary with the 'subgroup_by' values as keys and the documents as values.
+            - subgroup_ordered_keys: A list of 'subgroup_by' values in the order they were encountered
+        """
+        subgroup_ordered_keys = []
+        document_subgroups: Dict[str, Dict] = defaultdict(lambda: defaultdict(list))
+        if self.subgroup_by:
+            for key, value in document_groups.items():
+                for doc in value:
+                    if self.subgroup_by in doc.meta:
+                        document_subgroups[key][str(doc.meta[self.subgroup_by])].append(doc)
+                        if str(doc.meta[self.subgroup_by]) not in subgroup_ordered_keys:
+                            subgroup_ordered_keys.append(str(doc.meta[self.subgroup_by]))
+                    else:
+                        document_subgroups[key]["no_subgroup"].append(doc)
+                        if "no_subgroup" not in subgroup_ordered_keys:
+                            subgroup_ordered_keys.append("no_subgroup")
+        return document_subgroups, subgroup_ordered_keys
+
+    def _sort_docs(
+        self,
+        document_groups: Dict[str, List],
+        document_subgroups: Dict[str, Dict],
+        no_group: List,
+        ordered_keys: List,
+        subgroup_ordered_keys: List,
+    ):  # pylint: disable=too-many-positional-arguments
```

Comment:
this could become similar to the following:

```python
    def _merge_and_sort(self, groups: Dict[str, List[Document]], 
                       subgroups: Dict[str, Dict[str, List[Document]]]) -> List[Document]:
        result: List[Document] = []
        
        if subgroups:
            for group_docs in subgroups.values():
                for docs in group_docs.values():
                    result.extend(sorted(docs, key=lambda d: d.meta.get(self.sort_docs_by, float("inf"))) 
                                if self.sort_docs_by else docs)
            return result
        
        for docs in groups.values():
            result.extend(sorted(docs, key=lambda d: d.meta.get(self.sort_docs_by, float("inf"))) 
                        if self.sort_docs_by else docs)
            
        return result
```

Reply:
this sorting doesn't handle the ordering_keys